### ISSUE TITLE
Support scanning single packages

### ISSUE THREAD

**Issue by dmuhamedagic** [anchore/syft#1109]
Capability to scan single packages would be great. syft can already generate a catalogue of packages from an image or a directory: I assume that scanning only a package then shouldn't be too complicated.
With this feature it would be possible to skip generating SBOMs which have already been done provided they got stored in some repository.

**Comment by kzantow** [comment]
There are some different ways you can get individual targets, such as using the `file:` or `dir:` prefix. If you're looking to just catalog a single file, just use the `file:<file-path>` cataloger. 

Could you explain more what you are looking for here? Java JARs? Python files? etc.?

**Comment by sej7278** [comment]
RPM's would be nice, lets face it most of these formats are just zips/tarballs so perhaps an "archive" cataloger?

**Comment by kzantow** [comment]
You should be able to scan an RPM file directly using the `file:` prefix

**Comment by sej7278** [comment]
well what's odd is that only the syft-json output format seems to show details of the files inside the rpm, every other format just shows the rpm metadata.

**Comment by kzantow** [comment]
@sej7278 do you happen to have an example container or RPM file you're using that you could point to?

**Comment by sej7278** [comment]
yup, reproducer for you (thanks for the help btw @kzantow) first one gives me 231kb worth of info (per/python/shell scripts, makefiles etc.) second gives me about 4kb :

```bash
cd /tmp
curl -O https://repo.almalinux.org/almalinux/9.1/BaseOS/x86_64/os/Packages/bash-completion-2.11-4.el9.noarch.rpm

SYFT_FILE_METADATA_CATALOGER_ENABLED=true SYFT_FILE_METADATA_DIGESTS=sha1,sha256 \
   syft packages . -v -o syft-json | jq

SYFT_FILE_METADATA_CATALOGER_ENABLED=true SYFT_FILE_METADATA_DIGESTS=sha1,sha256 \
    syft packages . -v -o spdx-json | jq

```

i guess it could be a limitation of the non-sfyt formats, as converting from syft-json to spdx-json gives me similar results.

**Comment by kzantow** [comment]
Thanks @sej7278 from what I see here, Syft is properly determining the package correctly and outputting it in all formats, e.g.:
```
syft file:bash-completion-2.11-4.el9.noarch.rpm                                   
 ✔ Indexed bash-completion-2.11-4.el9.noarch.rpm 
 ✔ Cataloged packages      [1 packages]
NAME             VERSION       TYPE 
bash-completion  1:2.11-4.el9  rpm 
```

What you're seeing in Syft JSON format is the _metadata_:
```json
   "metadata": {
    ...
    "files": [
     {
      "path": "/etc/bash_completion.d",
      "mode": 16877,
      "size": 0,
      "digest": {
       "algorithm": "",
       "value": ""
      },
      "userName": "root",
      "groupName": "root",
      "flags": ""
     },
```

These are not files Syft has scanned, so they do not show up in SPDX files output, but rather only information Syft has surfaced from the RPM file's metadata.

I think what you're asking for is Syft to extract these and treat them as real files? The problem is, these haven't been found on the filesystem. If you _installed_ the RPM, and scanned the whole filesystem, these _would_ be found. I'm not entirely sure the right thing to do here, but I don't think surfacing these as files that Syft found is the right thing. Could you clarify if I've misunderstood something?

**Comment by sej7278** [comment]
ah ok, i assumed as there was a sha256sum it was extracting the files from the rpm.

so seems the best way to achieve what i need (don't want to install+scan) is a file scan of the source files and the resulting rpm's, so like pointing it at [mock's](https://github.com/rpm-software-management/mock/) builddir:

```
├── BUILD - prepared source code
├── BUILDROOT - compiled files in fakeroot e.g. /usr and /etc directories
├── originals - source SRPM
├── RPMS - generated RPM's
├── SOURCES - .patch files, tarball, makefiles
├── SPECS - .spec file
└── SRPMS - rebuilt SRPM
```

that way it gets the metadata from the rpm's as well as scans the sourcecode.

```
SYFT_FILE_METADATA_CATALOGER_ENABLED=true SYFT_FILE_METADATA_DIGESTS=sha1,sha256 \
    syft packages . --exclude **/.git/** -o spdx-json --file ~/spdx-json.json
```


cyclonedx doesn't seem to gather much other than the rpm stuff, but spdx/syft formats are fine.

**Comment by kzantow** [comment]
This sounds like a good idea, we've been thinking about extracting archives for a while and this falls directly into that category. We'll add this to the backlog.
